cpp/src/arrow/compute/light_array_internal.cc

```diff
@@ -235,6 +237,7 @@ void ResizableArrayData::Clear(bool release_buffers) {
 
 Status ResizableArrayData::ResizeFixedLengthBuffers(int num_rows_new) {
   ARROW_DCHECK(num_rows_new >= 0);
+  ARROW_DCHECK(data_type_ != nullptr);
```

Comment:
Why do we need this (and several other similar) check? I think we can assume that these functions will always be called after an `Init` call, where `data_type_` would be set to a non-null.

Rather we should add such a check in `Init` instead: the first use of passed `data_type` in `Init` is:
https://github.com/apache/arrow/pull/43760/files/990e9a15042a31fa8e9cfc07cb62da65cd11092e..87b5d061a8fe2c027b29335b07d2c66969688a23#diff-dca2a0b71d9a10c8634649df065d2631cadf93e3a05eefa5b83ae9d55348b63fR219
will give NPE.

Reply:
> Why do we need this (and several other similar) check? I think we can assume that these functions will always be called after an Init call, where data_type_ would be set to a non-null.

I just `DCHECK(data_type_)` to `DCHECK it's initialized`

Reply:
It just feels weird to ensure having been initialized by checking a not-so-relative member `data_type_`. But OK I get it. If so, please also check the passed-in `data_type` of `Init` being non-null - you can see that we don't have any guard on its nullability, so we may end up with NPE.

Reply:
> It just feels weird to ensure having been initialized by checking a not-so-relative member data_type_.

Since we don't have initialized, lol

Reply:
Added in https://github.com/apache/arrow/pull/43760/commits/b2223ffecf5c7c004a8be57cad7f0b7f9605623e

Reply:
I agree with @zanmato1984 that this doesn't seem terribly useful.